modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php

```diff
@@ -0,0 +1,295 @@
+<?php
+/**
+ * Image Loading Optimization: ILO_Grouped_URL_Metrics class
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * URL metrics grouped by minimum viewport width for the provided breakpoints.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class ILO_Grouped_URL_Metrics {
+
+	/**
+	 * URL metrics grouped by minimum viewport width for the provided breakpoints.
+	 *
+	 * @var array<int, ILO_URL_Metric[]>
+	 */
+	private $groups;
+
+	/**
+	 * Breakpoints in max widths.
+	 *
+	 * @var int[]
+	 */
+	private $breakpoints;
+
+	/**
+	 * Sample size for URL metrics for a given breakpoint.
+	 *
+	 * @var int
+	 */
+	private $sample_size;
+
+	/**
+	 * Freshness age (TTL) for a given URL metric.
+	 *
+	 * @var int
+	 */
+	private $freshness_ttl;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param ILO_URL_Metric[] $url_metrics   URL metrics.
+	 * @param int[]            $breakpoints   Breakpoints in max widths.
+	 * @param int              $sample_size   Sample size for the maximum number of viewports in a group between breakpoints.
+	 * @param int              $freshness_ttl Freshness age (TTL) for a given URL metric.
+	 */
+	public function __construct( array $url_metrics, array $breakpoints, int $sample_size, int $freshness_ttl ) {
+		$this->breakpoints   = $breakpoints;
+		$this->sample_size   = $sample_size;
+		$this->freshness_ttl = $freshness_ttl;
+		$this->groups        = $this->group_url_metrics_by_breakpoint( $url_metrics );
+	}
+
+	/**
+	 * Gets grouped keyed by the minimum viewport width.
+	 *
+	 * @return array<int, ILO_URL_Metric[]> Groups.
+	 */
+	public function get_groups(): array {
+		return $this->groups;
+	}
+
+	/**
+	 * Gets minimum viewport widths for the groups of URL metrics divided by the breakpoints.
+	 *
+	 * @return int[]
+	 */
+	public function get_minimum_viewport_widths(): array {
+		return array_keys( $this->groups );
+	}
+
+	/**
+	 * Unshifts a new URL metric, potentially pushing out older URL metrics when exceeding the sample size.
```

Comment:
I forgot, is it actually relevant that this is "unshifted"? If so, we may still want to clarify the documentation here. Potentially say something like that the metric will be prepended to the correct group, and the last entry from that group will be discarded if the sample size is exceeded.

Reply:
I forgot to update the phpdoc when I renamed the method from `unshift` to `add`. The fact that it does an unshift is a non-important implementation detail which should be hidden.

```suggestion
	 * Adds a new URL metric to a group.
	 *
	 * Once a group reaches the sample size, the oldest URL metric is pushed out.
```